src/main/java/org/jenkinsci/plugins/workflow/support/actions/LogStorageAction.java

```diff
@@ -0,0 +1,159 @@
+import hudson.model.AbstractBuild;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.io.Serializable;
+import java.io.UnsupportedEncodingException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.apache.commons.jelly.XMLOutput;
+import org.jenkinsci.plugins.workflow.actions.FlowNodeAction;
+import org.jenkinsci.plugins.workflow.actions.LogAction;
+import org.jenkinsci.plugins.workflow.actions.PersistentAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.log.LogStorage;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * A marker for a node which had some log text using {@link LogStorage#nodeListener}.
+ */
+@Restricted(NoExternalUse.class) // for use from DefaultStepContext only
+public class LogStorageAction extends LogAction implements FlowNodeAction, PersistentAction {
+
+    private static final Logger LOGGER = Logger.getLogger(LogStorageAction.class.getName());
+
+    public transient FlowNode node;
+
+    private LogStorageAction(FlowNode node) {
+        if (!node.isActive()) {
+            throw new IllegalStateException("cannot start writing logs to a finished node " + node);
+        }
+        this.node = node;
+    }
+
+    @Override public void onLoad(FlowNode node) {
+        this.node = node;
+    }
+
+    @Override public AnnotatedLargeText<? extends FlowNode> getLogText() {
+        FlowExecutionOwner owner = node.getExecution().getOwner();
+        return LogStorage.of(owner).stepLog(node, !node.isActive());
+    }
+
+    /**
+     * Used from <tt>console.jelly</tt> to write annotated log to the given output.
+     */
+    @Restricted(DoNotUse.class) // Jelly
+    public void writeLogTo(long offset, XMLOutput out) throws IOException {
+        getLogText().writeHtmlTo(offset, out.asWriter());
+    }
+
+    /**
+     * Creates a sink to print output from a step.
+     * Will use {@link LogActionImpl} if necessary.
+     * @param node a node which wishes to print output
+     * @param filter an optional log filter (must be {@link Serializable})
+     * @return a stream
+     */
+    @SuppressWarnings("deprecation") // LogActionImpl here for backward compatibility
+    public static @Nonnull TaskListener listenerFor(@Nonnull FlowNode node, @CheckForNull ConsoleLogFilter filter) throws IOException, InterruptedException {
+        FlowExecutionOwner owner = node.getExecution().getOwner();
+        if (LogActionImpl.isOld(owner) || node.getAction(LogActionImpl.class) != null) {
+            return LogActionImpl.stream(node, filter);
```

Comment:
Probably the `!FlowNode.active` closing logic from `LogActionImpl` should be moved here—for some implementations it may be significant (though not for `FileLogStorage`). Also rather than `PrintStream.close` it should check for `TaskListener & AutoCloseable`. And the expected close behavior for both the overall and step logs should be documented in `LogStorage`.

Reply:
(resolved)